pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -54,29 +55,61 @@ type AmbientIndex struct {
 	byService map[string][]*model.WorkloadInfo
 	// byPod indexes by Pod IP address.
 	byPod map[string]*model.WorkloadInfo
+	// services are indexed by the network/clusterIP
+	// TODO: GregHanson index by service hostname+port?
+	services map[string]*workloadapi.Service
 
 	// Map of ServiceAccount -> IP
 	// TODO: currently, this is derived from pods. To be agnostic to the implementation,
 	// we should actually be looking at Gateway.status.addresses.
 	// This may be an external address (possibly even a DNS name we need to resolve), an arbitrary IP,
 	// or a reference to a service.
 	// If its a reference to a Service then we can find the underlying pods in that service, as an optimization.
-	waypoints map[model.WaypointScope]sets.String
+	waypoints map[model.WaypointScope]*workloadapi.GatewayAddress
 
 	// serviceVipIndex maintains an index of VIP -> Service
 	serviceVipIndex *kclient.Index[string, *v1.Service]
 }
 
 // Lookup finds a given IP address.
-func (a *AmbientIndex) Lookup(ip string) []*model.WorkloadInfo {
+func (a *AmbientIndex) Lookup(key string) []*model.AddressInfo {
 	a.mu.RLock()
 	defer a.mu.RUnlock()
 	// First look at pod...
-	if p, f := a.byPod[ip]; f {
-		return []*model.WorkloadInfo{p}
+	if p, f := a.byPod[key]; f {
+		addr := &workloadapi.Address{
+			Type: &workloadapi.Address_Workload{
+				Workload: p.Workload,
+			},
+		}
+		return []*model.AddressInfo{{Address: addr}}
 	}
+	
 	// Fallback to service. Note: these IP ranges should be non-overlapping
-	return a.byService[ip]
+	// cannot distunagish between Service lookup and Workloads for a Service lookup
+	// since the same newtork/vip key is used.  Both Service and Workloads are
+	// returned in this case
+	// TODO any service to exclude in the future? for example with affinity, 
+	// per-node, etc - where simply load balancing will break?
+	res := []*model.AddressInfo{}
+	for _, wl := range a.byService[key] {
+		addr := &workloadapi.Address{
+			Type: &workloadapi.Address_Workload{
+				Workload: wl.Workload,
+			},
+		}
+		res = append(res, &model.AddressInfo{Address: addr})
+	}
+	if s, exists := a.services[key]; exists {
```

Comment:
if `len(res) > 0` before we enter this loop and we don't add any services, we may want a warning log line? i don't think we'd expect to get workloads that can't be reached by any services

Reply:
im walking this back, end of day vibes or something. i don't think we want this; could be very reasonable to lookup workloads and have no services that select them

Reply:
i think the more bizarre and weird case is if we add both a workload and service for the same IP.. that collision could be confusing for users depending on which we prefer

Reply:
by default the pod is always returned.  If something is found in the byPod lookup, we return and never check for any possible service collisions.  I could add a comment to the method that that is the behavior/precedence in the event of a collision